### PR TITLE
Remove buttons and context from action manager, fix multi-viewer

### LINKED ISSUE #2897: Viewer buttons don't work as expected when having two napari viewers in a single program
## 🐛 Bug

Hi,

Since napari 0.4.8, the buttons that by default are in the lower left corner don't work as expected when having two napari viewers in a single program. More specifically, when clicking one of the lower left corner buttons in the napari viewer that was added first, the button's action is executed on the other viewer. This seems to be related to the new action manager features that were added in 0.4.8.

I'm not sure if this is actually a bug or if napari simply isn't meant to be used like this, but until 0.4.8 I hadn't experienced any issues, so it would be nice if it could continue to work when used like this.

Demo

## To Reproduce

1. Run the following minimal working example with napari 0.4.8 or later:
```python
import napari
from qtpy import QtWidgets
from skimage import data


class Window(QtWidgets.QWidget):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.napari1 = napari.Viewer(show=False)
        self.napari1.add_image(data.cat())
        self.napari2 = napari.Viewer(show=False)
        self.napari2.add_image(data.astronaut())

        self.setLayout(QtWidgets.QHBoxLayout())
        self.layout().addWidget(self.napari1.window._qt_window)
        self.layout().addWidget(self.napari2.window._qt_window)


app = QtWidgets.QApplication([])
window = Window()
window.show()
app.exec_()
```
2. Click the console button in the lower left corner of the napari viewer that displays the cat image.

The console will show up in the napari viewer displaying the astronaut image, instead of in the one displaying the cat image.

The same issue applies to the other buttons in the lower left corner.

## Environment

```
napari: 0.4.9
Platform: Windows-10-10.0.19041-SP0
Python: 3.9.4 (default, Apr 9 2021, 11:43:21) [MSC v.1916 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.19.3
SciPy: 1.6.3
Dask: 2021.05.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 465.89
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0

Plugins:
- console: 0.0.3
- scikit-image: 0.4.9
- svg: 0.1.5
```

## Additional context

The keyboard shortcuts for the buttons work as expected.

### LINKED ISSUE #2977: Unable to use multiple windows
## 🐛 Bug

My problem is that when I open multiple napari windows, the context of the window changes completely to the second window. One problem with this, is that when I try to change to 3D mode in the first window, nothing happens, and I get an error:
```
WARNING: Traceback (most recent call last):
  File "C:/Users/Tejfel/AppData/Roaming/JetBrains/PyCharmCE2021.1/scratches/napari_bug_test.py", line 13, in <module>
    napari.run()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\_qt\qt_event_loop.py", line 376, in run
    app.exec_()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\action_manager.py", line 55, in <lambda>
    self.command, context
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\action_manager.py", line 43, in call_with_context
    return function(**ctx)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\components\_viewer_key_bindings.py", line 48, in toggle_ndisplay
    viewer.dims.ndisplay = 3
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\evented_model.py", line 155, in __setattr__
    getattr(self.events, name)(value=after)  # emit event
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\event.py", line 573, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\event.py", line 593, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\_vispy\vispy_camera.py", line 138, in _on_ndisplay_change
    self._view.camera = self._3D_camera
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\frozen.py", line 17, in __setattr__
    object.__setattr__(self, key, value)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\widgets\viewbox.py", line 103, in camera
    cam.parent = self.scene
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\node.py", line 200, in parent
    parent._add_child(self)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\node.py", line 214, in _add_child
    self.events.children_change(added=node)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\canvas.py", line 333, in _update_scenegraph
    self.update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\canvas.py", line 200, in update
    super(SceneCanvas, self).update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\app\canvas.py", line 442, in update
    self._backend._vispy_update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\app\backends\_qt.py", line 427, in _vispy_update
    self.update()
RuntimeError: wrapped C/C++ object of type CanvasBackendDesktop has been deleted
```
You can also see this happening, if you create two windows, close the second one, and try to access `viewer` in the console of the first window. Actually the second window's `viewer` gets returned.
I guess  you didn't create napari to utilize multiple windows simultaneously, so this might be a feature request rather than a bug report, but it would help if I could open a second window.

Actually I can manually set the `viewer` back to the original one programmatically from the console, but this won't help with the ndisplay problem.

Any help would be highly appreciated!
## To Reproduce

1. run code:
```
import napari
from magicgui import magicgui
from napari.layers import Image
import numpy as np

@magicgui(call_button="New window")
def napari_new_viewer():
    viewer2 = napari.Viewer()
    viewer2.add_layer(Image(np.ones((10, 10)), name="image2"))
viewer1 = napari.Viewer()
viewer1.add_layer(Image(np.ones((10, 10)), name="image1"))
viewer1.window.add_dock_widget(napari_new_viewer)
napari.run()
```
2. Click "New window" in the added widget
3. Close second window
4. Run `viewer.layers` in the IPython console
    * "image2" will be returned instead of "image1"
5. Click Toggle ndisplay
    * an error will occure, and it won't change to 3D view

## Expected behavior

For me two options would work well:
a) The context changes when the second window is created, and changes back, when it is closed
b) The app detects that a window (and a viewer) was already created, so it doesn't do anything in the background
In the second window I only need to display the image, segment it, and check it both in 2D and 3D, so a minimal version of napari would be sufficient.

## Environment
napari: 0.4.10
Platform: Windows-10-10.0.19041-SP0
Python: 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 15 2019, 00:11:34) [MSC v.1916 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.19.5
SciPy: 1.4.1
Dask: 2021.06.2
VisPy: 0.6.6

OpenGL:
- GL version: 4.5.0 - Build 25.20.100.6577
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- scikit-image: 0.4.10
- svg: 0.1.5

## Additional context
Actually what I'm trying to do is to allow the user to select a part of the original nD file, and segment it as a separate unit. This would allow the user to visually check it in 3D without loading everything into memory. If it is impossible to use multiple windows simultaneously, than I will try to display the selected part in the original window, and hide the original data, but it would be much nicer and natural to open a second window.

### PR DESCRIPTION
**Pull request by tlambert03**
# Description

closes #2914, should fix #2977 and fix #2897

This "relieves" the action manager from the duty of remembering buttons, by adding a "shortcut_changed" event to which buttons can subscribe and update their tooltips accordingly.  (That really was the main thing that we needed from the action manager controlling buttons).

This also removes the context object from the action manager, which was also partially causing our difficulties with multiple viewers.  The only thing we currently needed context for is to get the current viewer, or the active layer, so I just added a small wrapper function to all of the viewer/layer actions that gets the current viewer/layer.  This dependency injection pattern could be elaborated in the (new) `ActionManager.trigger` method, which could inspect the type annotations of parameters and provide standard objects based on rules (kinda like magicguidoes).

**update:**
I added a more general `inject_napari_dependencies` function to this PR.  It's a function that takes a callable and returns a callable.  If the callable has any parameters that are annotated as certain napari types (like `Viewer` or `Layer`), those parameters are optional at calltime, otherwise they will be injected as the current viewer, or the active layer, etc.

```py
>>> def f(viewer: Viewer): ...
>>> inspect.signature(f)
<Signature (x: 'Viewer')>

>>> f2 = inject_napari_dependencies(f)
>>> inspect.signature(f2)
<Signature (x: typing.Optional[napari.viewer.Viewer] = None)>

# if f2 is called without x, the current_viewer will be provided for x
```

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
add a vispy warning ignore, so that tests can run, but this should be removed after https://github.com/napari/napari/issues/3693 is fixed

**Comment by tlambert03** [comment]
This is ready to go.  just the mac screenshot leak failure now

File: napari/utils/action_manager.py [hunk 1]
```diff
@@ -175,16 +79,14 @@ class ActionManager:
     def __init__(self):
         # map associating a name/id with a Comm
         self._actions: Dict[str, Action] = {}
-        self._buttons: Dict[str, Set[ButtonWrapper]] = defaultdict(set)
+        # self._buttons: Dict[str, Set[ButtonWrapper]] = defaultdict(set)
```
Comment (tlambert03) [inline napari/utils/action_manager.py]:
```suggestion
```

**Review by sofroniewn — APPROVED** [review]
Changes all seem fine to me, but maybe @Carreau and @jni can give this a pass before merge

File: napari/_qt/_tests/test_qt_viewer.py [hunk 2]
```diff
@@ -486,7 +486,7 @@ def test_leaks_labels(qtbot, make_napari_viewer):
     )
     dr = weakref.ref(lr().data)
     viewer.layers.clear()
-    qtbot.wait(100)
+    qtbot.wait(200)  # TODO: figure out exactly what we're waiting for
```
Comment (jni) [inline napari/_qt/_tests/test_qt_viewer.py:501]: 😂

File: napari/_tests/test_multiple_viewers.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from napari import Viewer
+
+
+def test_multi_viewers_dont_clash(qapp):
+    v1 = Viewer(show=False, title='v1')
+    v2 = Viewer(show=False, title='v2')
+    assert not v1.grid.enabled
+    assert not v2.grid.enabled
+
+    v1.window.activate()  # a click would do this in the actual gui
```
Comment (jni) [inline napari/_tests/test_multiple_viewers.py:10]: ❤️

File: napari/components/_viewer_key_bindings.py [hunk 4]
```diff
@@ -41,11 +47,8 @@ def reset_scroll_progress(viewer):
 
 
 @register_viewer_action(trans._("Toggle ndisplay."))
-def toggle_ndisplay(viewer):
-    if viewer.dims.ndisplay == 3:
-        viewer.dims.ndisplay = 2
-    else:
-        viewer.dims.ndisplay = 3
+def toggle_ndisplay(viewer: Viewer):
+    viewer.dims.ndisplay = 2 if viewer.dims.ndisplay == 3 else 3
```
Comment (jni) [inline napari/components/_viewer_key_bindings.py]: Are we golfing? 😜 

```suggestion
    viewer.dims.ndisplay = 2 + (viewer.dims.ndisplay == 2)
```

File: napari/utils/action_manager.py [hunk 5]
```diff
@@ -1,45 +1,35 @@
...
 
 import warnings
-import weakref
 from collections import defaultdict
 from dataclasses import dataclass
-from inspect import signature
-from typing import TYPE_CHECKING, Callable, Dict, List, Set, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Set, Union
 
+from .._vendor.cpython.functools import cached_property
```
Comment (jni) [inline napari/utils/action_manager.py:8]: Wait what? When did this happen? 😂 But ok, seems worthwhile.

**Review by jni — APPROVED** [review]
Love this, can't wait to have this out. Thanks @tlambert03!!!

File: napari/components/_viewer_key_bindings.py (hunk 4 above)
Comment (jni) [reply napari/components/_viewer_key_bindings.py]: (Note, I'm not advocating for one approach over the other here, happy to ignore this, I just couldn't resist. =)

File: napari/utils/action_manager.py (hunk 5 above)
Comment (tlambert03) [reply napari/utils/action_manager.py:8]: A month ago or so, it's a very useful decorator.  We can drop when we drop support for py3.7

File: napari/components/_viewer_key_bindings.py (hunk 4 above)
Comment (tlambert03) [reply napari/components/_viewer_key_bindings.py]: That's just an auto refactor of the existing code.  Happy to use yours too

**Comment by sofroniewn** [comment]
@tlambert03 now many to the "slow" usage has been dropped can you update from main and then if all tests pass we can merge

**Comment by tlambert03** [comment]
@Carreau, this is mostly just waiting on a comment from you.  would prefer not to merge without you taking a quick glance

**Comment by Czaki** [comment]
Why `QtStateButton` was removed without deprecation? This was a part of public `napari.qt` API.

**Comment by tlambert03** [comment]
Sorry about that! can you open an issue/PR to add it back?  Do you actually use it (you're probably the only one).  

we don't need it anymore, so you might consider just copying that code and using it internally in PartSeg ...

File: napari/_qt/_tests/test_qt_viewer.py (hunk 2 above)
Comment (Czaki) [reply napari/_qt/_tests/test_qt_viewer.py:501]: on cleaning of qt objects

Comment (tlambert03) [reply napari/_qt/_tests/test_qt_viewer.py:501]: yes, I understand that... but we should pretty much always be using `qtbot.waitSignal()` or something more specific.  using `wait(some_arbitrary_delay)` is too vague.  we really don't have much async stuff going on here, so we should be able to specify what we're waiting for, and avoid sporadic test failures

Comment (Czaki) [reply napari/_qt/_tests/test_qt_viewer.py:501]: [`destroyed()`](https://doc.qt.io/qt-5/qobject.html#destroyed) ? But which object?

Comment (tlambert03) [reply napari/_qt/_tests/test_qt_viewer.py:501]: Exactly :). That's what the comment was trying to say.  exacty what?

**Pull request merged**